docs/DeveloperGuide.adoc

```diff
@@ -267,6 +268,96 @@ are applied when restoring `**THRIFT**` data from the data file.
+
+Step 1. Application is launched with empty transaction list. Both `UndoStack` and `RedoStack` are empty upon initializing `**PastUndoableCommands**`.
+
+image::UndoAndRedo/initialUndoRedo.PNG[333, 300]
+
+Step 2. The user executes *add_expense n/Laksa v/4* to add laksa to the `**TransactionList**`.
+The `LogicManager` invokes the method `Model#keepTrackCommands(Undoable undoable)` to keep track of the `AddExpenseCommand` object
+that was used to add laksa. This `AddExpenseCommand` object is pushed into `UndoStack`.
+
+image::UndoAndRedo/addExpense.PNG[333, 300]
+
+Step 3. The user executes *update i/1 v/3.5* to update the cost of the laksa to $3.50.
+The `LogicManager` calls the method `Model#keepTrackCommands(Undoable undoable)` to keep track of the `UpdateCommand` object
+that was used to update the cost of the laksa. This `UpdateCommand` object is pushed into `UndoStack`.
+
+image::UndoAndRedo/updateExpense.PNG[346, 300]
+
+Step 4. The user executes *undo* to revert the cost of the laksa back to $4.00.
+The `UndoCommand` invokes the method `Model#getPreviousUndoableCommand()` to retrieve the `UpdateCommand` object.
+Before returning the `UpdateCommand` object back to the callee, `UpdateCommand` is popped from `UndoStack` and pushes into `RedoStack`.
+
+image::UndoAndRedo/undo.PNG[315, 300]
+
+Step 5. The user executes *undo* to remove laksa from `**TransactionList**`.
+The `UndoCommand` invokes the method `Model#getPreviousUndoableCommand()` to retrieve the `AddExpenseCommand` object.
+Before returning the `AddExpenseCommand` back to the callee, `AddExpenseCommand` is popped from `UndoStack` and pushes into `RedoStack`.
+
+image::UndoAndRedo/undo2.PNG[315, 300]
+
+The sequence diagram shows how *undo* works for *add_expense* command:
+
+.Sequence diagram of how *undo* command is applied at the back-end
+image::UndoAndRedo/UndoAddExpenseSequenceDiagram.png[]
+
+Step 6. The user executes *redo* to add laksa back to `TransactionList`.
+The `RedoCommand` invokes the method `Model#getUndoneCommand()` to retrieve the `AddExpenseCommand` object.
+Before returning the `AddExpenseCommand` back to the callee, `AddExpenseCommand` was pop from `RedoStack` and push into `UndoStack`.
+
+image::UndoAndRedo/redo.PNG[310, 300]
+
+NOTE: All commands in `RedoStack` are purged when user executes another undoable command.
+
+The sequence diagram shows how *redo* works for *add_expense* command:
+
+.Sequence diagram of how *redo* command is applied at the back-end
+image::UndoAndRedo/RedoAddExpenseSequenceDiagram.png[]
+
+The following activity diagram shows what happen if user executes command:
+
+.Activity diagram of the command execution.
+image::UndoAndRedo/RunCommandActivityDiagrams.png[401, 600]
+
+==== Design considerations
+
+===== Aspect: Data structure used to keep track of undoable command
+
+* **Alternative 1 (current choice):** Use two stacks. One stack stores the command to undo, the other stack stores the command to redo.
```

Comment:
... the command to undo **while** the ...

Reply:
Resolved